[DETAIL] Table of contents links break due to duplicate auto-generated heading IDs across prose blocks

> [!NOTE]
> From https://app.detail.dev/oss/opengovsg/isomer/bug_9b56534b-b9d6-4106-b1f9-ec53b01d301c
 
# Summary
- **Context**: The `getDigestFromText` function generates MD5 hash IDs used for table of contents anchor links and block identifiers.
- **Bug**: IDs for prose level-2 headings are generated without the outer block index, causing duplicate IDs when multiple prose blocks contain identical headings at the same position.
- **Actual vs. expected**: Actual: identical level-2 headings in different prose blocks can receive the same `id`, producing duplicate HTML IDs and TOC links that both point to the first heading. Expected: each heading should receive a unique `id` so each TOC entry navigates to its corresponding heading.
- **Impact**: Invalid HTML (duplicate IDs), broken table of contents anchor links that only navigate to the first matching element, and accessibility violations.

# Code with Bug
```typescript
// packages/components/src/utils/getTransformedPageContent.ts
const anchorId = getDigestFromText(
  `${JSON.stringify(component)}_${componentIndex}`,
) // <-- BUG 🔴 missing outer prose-block index, so headings in different blocks can collide
```

# Explanation
`getTransformedPageContent` auto-generates IDs only for level-2 headings that don’t already have an `id`. The hash input includes only the serialized heading node and its position within a single prose block (`componentIndex`). If multiple prose blocks contain the same heading structure at the same inner index (e.g., both blocks start with an `h2` “Introduction”), the MD5 inputs are identical, so the IDs collide. `getTableOfContents` then builds `anchorLink: "#" + component.attrs.id`, so multiple TOC entries can end up linking to the same DOM element (the first matching ID).

## Codebase Inconsistency
Complex blocks already include the outer block index when generating IDs, avoiding collisions:
```typescript
// packages/components/src/utils/getTransformedPageContent.ts
return {
  ...block,
  id: getDigestFromText(`${JSON.stringify(block)}_${index}`),
}
```

# Failing Test
```typescript
// packages/components/src/utils/__tests__/getTransformedPageContent.test.ts
it("generates unique IDs for identical headings in different prose blocks", () => {
  const content: IsomerComponent[] = [
    {
      type: "prose",
      content: [
        {
          type: "heading",
          attrs: { level: 2 },
          content: [{ type: "text", text: "Introduction" }],
        },
      ],
    },
    {
      type: "prose",
      content: [
        {
          type: "heading",
          attrs: { level: 2 },
          content: [{ type: "text", text: "Introduction" }],
        },
      ],
    },
  ]
  const transformed = getTransformedPageContent(content)
  const heading1 = transformed[0]?.content?.[0]
  const heading2 = transformed[1]?.content?.[0]
  expect(heading1?.attrs?.id).toMatch(idPattern)
  expect(heading2?.attrs?.id).toMatch(idPattern)
  expect(heading1?.attrs?.id).not.toBe(heading2?.attrs?.id)
})
```

Test output:
```
FAIL  src/utils/__tests__/getTransformedPageContent.test.ts > getTransformedPageContent > generates unique IDs for identical headings in different prose blocks
AssertionError: expected '9c479136a83119bff5c9bdc9b6bb9424' not to be '9c479136a83119bff5c9bdc9b6bb9424' // Object.is equality
```

# Recommended Fix
```typescript
// packages/components/src/utils/getTransformedPageContent.ts
const anchorId = getDigestFromText(
  `${JSON.stringify(component)}_${index}_${componentIndex}`,
)
```

# History
This bug was introduced in commit e0fd6bb3.

--- a/packages/components/src/utils/__tests__/getTransformedPageContent.test.ts
+++ b/packages/components/src/utils/__tests__/getTransformedPageContent.test.ts
@@ -89,7 +89,7 @@ describe("getTransformedPageContent", () => {
     }
   })
 
-  it("generates unique IDs for identical headings in different prose blocks", () => {
+  it("generates unique IDs for identical headings in 2 prose blocks", () => {
     // Arrange
     const proseBlock: IsomerComponent = {
       type: "prose",
@@ -115,6 +115,71 @@ describe("getTransformedPageContent", () => {
     })
 
     expect(ids[0]).toBe(getDigestFromText(seed)) // preserve legacy seed for the first heading
+    expect(ids[1]).toBe(getDigestFromText(`${seed}_1`))
+    expect(ids[0]).not.toBe(ids[1])
+  })
+
+  it("generates unique IDs for identical headings in 3+ prose blocks", () => {
+    // Arrange
+    const proseBlock: IsomerComponent = {
+      type: "prose",
+      content: [
+        {
+          type: "heading",
+          attrs: { level: 2 },
+          content: [{ type: "text", text: "Introduction" }],
+        },
+      ],
+    }
+    const seed = `${JSON.stringify(proseBlock.content[0])}_0`
+    const content: IsomerComponent[] = [proseBlock, proseBlock, proseBlock]
+
+    // Act
+    const transformed = getTransformedPageContent(content)
+
+    // Assert
+    const ids = transformed.map((block) => {
+      if (block.type !== "prose" || !block.content) return undefined
+      const node = block.content[0]
+      return node?.type === "heading" ? node.attrs.id : undefined
+    })
+
+    expect(ids[0]).toBe(getDigestFromText(seed))
+    expect(ids[1]).toBe(getDigestFromText(`${seed}_1`))
+    expect(ids[2]).toBe(getDigestFromText(`${seed}_2`))
+    expect(new Set(ids).size).toBe(3)
+  })
+
+  it("preserves componentIndex in seed when heading is not first in prose block", () => {
+    // Arrange
+    const proseBlock: IsomerComponent = {
+      type: "prose",
+      content: [
+        {
+          type: "paragraph",
+          content: [{ type: "text", text: "Intro paragraph" }],
+        },
+        {
+          type: "heading",
+          attrs: { level: 2 },
+          content: [{ type: "text", text: "Introduction" }],
+        },
+      ],
+    }
+    const seed = `${JSON.stringify(proseBlock.content[1])}_1`
+    const content: IsomerComponent[] = [proseBlock, proseBlock]
+
+    // Act
+    const transformed = getTransformedPageContent(content)
+
+    // Assert
+    const ids = transformed.map((block) => {
+      if (block.type !== "prose" || !block.content) return undefined
+      const node = block.content[1]
+      return node?.type === "heading" ? node.attrs.id : undefined
+    })
+
+    expect(ids[0]).toBe(getDigestFromText(seed))
     expect(ids[1]).toBe(getDigestFromText(`${seed}_1`))
     expect(ids[0]).not.toBe(ids[1])
   })

--- a/packages/components/src/utils/__tests__/getTransformedPageContent.test.ts
+++ b/packages/components/src/utils/__tests__/getTransformedPageContent.test.ts
@@ -1,5 +1,6 @@
 import type { IsomerComponent } from "~/types"
 import { describe, expect, it } from "vitest"
+import { getDigestFromText } from "~/utils/getDigestFromText"
 import { getTransformedPageContent } from "~/utils/getTransformedPageContent"
 
 describe("getTransformedPageContent", () => {
@@ -100,6 +101,7 @@ describe("getTransformedPageContent", () => {
         },
       ],
     }
+    const seed = `${JSON.stringify(proseBlock.content[0])}_0`
     const content: IsomerComponent[] = [proseBlock, proseBlock]
 
     // Act
@@ -108,12 +110,12 @@ describe("getTransformedPageContent", () => {
     // Assert
     const ids = transformed.map((block) => {
       if (block.type !== "prose" || !block.content) return undefined
-      const heading = block.content[0]
-      return heading?.type === "heading" ? heading.attrs.id : undefined
+      const node = block.content[0]
+      return node?.type === "heading" ? node.attrs.id : undefined
     })
 
-    expect(ids[0]).toMatch(idPattern)
-    expect(ids[1]).toMatch(idPattern)
+    expect(ids[0]).toBe(getDigestFromText(seed)) // preserve legacy seed for the first heading
+    expect(ids[1]).toBe(getDigestFromText(`${seed}_1`))
     expect(ids[0]).not.toBe(ids[1])
   })
 

--- a/packages/components/src/utils/__tests__/getTransformedPageContent.test.ts
+++ b/packages/components/src/utils/__tests__/getTransformedPageContent.test.ts
@@ -88,6 +88,35 @@ describe("getTransformedPageContent", () => {
     }
   })
 
+  it("generates unique IDs for identical headings in different prose blocks", () => {
+    // Arrange
+    const proseBlock: IsomerComponent = {
+      type: "prose",
+      content: [
+        {
+          type: "heading",
+          attrs: { level: 2 },
+          content: [{ type: "text", text: "Introduction" }],
+        },
+      ],
+    }
+    const content: IsomerComponent[] = [proseBlock, proseBlock]
+
+    // Act
+    const transformed = getTransformedPageContent(content)
+
+    // Assert
+    const ids = transformed.map((block) => {
+      if (block.type !== "prose" || !block.content) return undefined
+      const heading = block.content[0]
+      return heading?.type === "heading" ? heading.attrs.id : undefined
+    })
+
+    expect(ids[0]).toMatch(idPattern)
+    expect(ids[1]).toMatch(idPattern)
+    expect(ids[0]).not.toBe(ids[1])
+  })
+
   it("adds ids to supported structured blocks", () => {
     // Arrange
     const content: IsomerComponent[] = [

--- a/packages/components/src/utils/getTransformedPageContent.ts
+++ b/packages/components/src/utils/getTransformedPageContent.ts
@@ -19,7 +19,7 @@ export const getTransformedPageContent = (
           ) {
             // generate a unique hash to auto-generate anchor links
             const anchorId = getDigestFromText(
-              `${JSON.stringify(component)}_${componentIndex}`,
+              `${JSON.stringify(component)}_${index}_${componentIndex}`,
             )
             const newAttrs = {
               ...component.attrs,

--- a/packages/components/src/utils/getTransformedPageContent.ts
+++ b/packages/components/src/utils/getTransformedPageContent.ts
@@ -20,7 +20,8 @@ export const getTransformedPageContent = (
             component.attrs.id === undefined
           ) {
             // generate a unique hash to auto-generate anchor links
-            // for backwards compatibility, use the same seed format for headings, but with an added index suffix only if there are multiple headings with the same content
+            // for backwards compatibility, use the seed format for headings,
+            // add index suffix only if for multiple headings with the same content
             const seed = `${JSON.stringify(component)}_${componentIndex}`
             let anchorId = getDigestFromText(seed)
             if (usedHeadingIds.has(anchorId)) {

--- a/packages/components/src/utils/getTransformedPageContent.ts
+++ b/packages/components/src/utils/getTransformedPageContent.ts
@@ -7,6 +7,8 @@ import { getDigestFromText } from "./getDigestFromText"
 export const getTransformedPageContent = (
   content: IsomerComponent[],
 ): IsomerComponent[] => {
+  const usedHeadingIds = new Set<string>()
+
   return content.map((block, index) => {
     if (block.type === "prose" && block.content) {
       return {
@@ -18,9 +20,14 @@ export const getTransformedPageContent = (
             component.attrs.id === undefined
           ) {
             // generate a unique hash to auto-generate anchor links
-            const anchorId = getDigestFromText(
-              `${JSON.stringify(component)}_${index}_${componentIndex}`,
-            )
+            // for backwards compatibility, use the same seed format for headings, but with an added collision count suffix only if there are multiple headings with the same content
+            const seed = `${JSON.stringify(component)}_${componentIndex}`
+            let anchorId = getDigestFromText(seed)
+            if (usedHeadingIds.has(anchorId)) {
+              anchorId = getDigestFromText(`${seed}_${index}`)
+            }
+            usedHeadingIds.add(anchorId)
+
             const newAttrs = {
               ...component.attrs,
               id: anchorId,

--- a/packages/components/src/utils/getTransformedPageContent.ts
+++ b/packages/components/src/utils/getTransformedPageContent.ts
@@ -20,7 +20,7 @@ export const getTransformedPageContent = (
             component.attrs.id === undefined
           ) {
             // generate a unique hash to auto-generate anchor links
-            // for backwards compatibility, use the same seed format for headings, but with an added collision count suffix only if there are multiple headings with the same content
+            // for backwards compatibility, use the same seed format for headings, but with an added index suffix only if there are multiple headings with the same content
             const seed = `${JSON.stringify(component)}_${componentIndex}`
             let anchorId = getDigestFromText(seed)
             if (usedHeadingIds.has(anchorId)) {

